eth/protocols/snap/sync.go

```diff
@@ -0,0 +1,718 @@
+// errors will be healed after the leaves are fully accumulated.
+func (s *Syncer) syncStorage(root common.Hash, account common.Hash, stroot common.Hash, peer *Peer, cancel chan struct{}) (bool, error) {
+	// For now simply iterate over the state iteratively
+	batch := s.db.NewBatch()
+
+	for {
+		// If the sync was cancelled, abort
+		select {
+		case <-cancel:
+			return false, nil
+		default:
+		}
+		// Generate a random request ID (clash probability is insignificant)
+		id := uint64(rand.Int63())
+
+		// Track the request and send it to the peer
+		s.lock.Lock()
+		s.storageReqs[peer.id] = &storageRequest{
+			id:     id,
+			root:   stroot,
+			origin: s.nextSlot,
+			cancel: cancel,
+		}
+		s.lock.Unlock()
+
+		if err := peer.RequestStorageRange(id, root, account, s.nextSlot, maxRequestSize); err != nil {
+			return false, err
+		}
+		s.storageRequests++
+
+		// Wait for the reply to arrive
+		res := <-s.storageResps
+		if res == nil {
+			return false, errors.New("unfulfilled request")
+		}
+		s.storageSynced += uint64(len(res.slots))
+		s.storageProofs += uint64(len(res.bounds))
+
+		// Hack
+		if res.nodes == nil {
+			break
+		}
+		// Persist every finalized trie node that's not on the boundary
+		it := res.nodes.NewIterator(nil, nil)
+		for it.Next() {
+			// Boundary nodes are not written, since they are incomplete
+			if _, ok := res.bounds[common.BytesToHash(it.Key())]; ok {
+				continue
+			}
+			// Node not a boundary, persist to disk
+			batch.Put(it.Key(), it.Value())
+			s.bloom.Add(it.Key())
+
+			s.storageNodes++
+			s.storageBytes += common.StorageSize(common.HashLength + len(it.Value()))
+		}
+		it.Release()
+
+		if err := batch.Write(); err != nil {
+			return false, err
+		}
+		batch.Reset()
+
+		// Storage range processed, step to the next chunk
+		log.Debug("Persisted range of storage slots", "account", account, "slot", res.last)
+		s.report(false)
+
+		// If the response contained all the data in one shot (no proofs), there
+		// is no reason to continue the sync, report immediate success.
+		if len(res.bounds) == 0 {
+			return true, nil
+		}
+		s.nextSlot = common.BigToHash(new(big.Int).Add(res.last.Big(), big.NewInt(1)))
+	}
+	return false, nil
+}
+
+// OnAccounts is a callback method to invoke when a range of accounts are
+// received from a remote peer.
+func (s *Syncer) OnAccounts(peer *Peer, id uint64, hashes []common.Hash, accounts [][]byte, proof [][]byte) error {
+	peer.Log().Trace("Delivering range of accounts", "hashes", len(hashes), "accounts", len(accounts), "proofs", len(proof))
+
+	// If the request is stale, discard it
+	s.lock.Lock()
+	req, ok := s.accountReqs[peer.id]
+	if !ok || req.id != id {
+		peer.Log().Warn("Unexpected account range packet")
+		s.lock.Unlock()
+		return nil
+	}
+	delete(s.accountReqs, peer.id)
+	root := s.root
+	s.lock.Unlock()
+
+	// If the response is unavailable snapshot, forward to the requester
```

Comment:
> If the response is unavailable snapshot

foo?

Reply:
If the response is "this snapshot is not available at the remote peer", forward to the requested.

Though I agree it needs a few takes to get it, I'll rephrase